pkg/controller/core/queue_controller.go

```diff
@@ -47,47 +55,124 @@ func NewQueueReconciler(queues *queue.Manager) *QueueReconciler {
+
+	pendingJobs, err := r.queues.Status(&queueObj)
+	if err != nil {
+		r.log.Error(err, "Failed to retrieve queue status")
+		return ctrl.Result{}, err
+	}
+
+	oldStatus := queueObj.Status
+	queueObj.Status.PendingWorkloads = int32(pendingJobs)
+	if !equality.Semantic.DeepEqual(oldStatus, queueObj.Status) {
+		err := r.client.Status().Update(ctx, &queueObj)
+		r.log.Error(err, "Failed to Update queue")
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
 	return ctrl.Result{}, nil
 }
 
 func (r *QueueReconciler) Create(e event.CreateEvent) bool {
-	q := e.Object.(*kueue.Queue)
+	q, match := e.Object.(*kueue.Queue)
+	if !match {
+		// No need to interact with the queue manager for other objects.
+		return true
+	}
 	log := r.log.WithValues("queue", klog.KObj(q))
 	log.V(2).Info("Queue create event")
 	ctx := logr.NewContext(context.Background(), log)
 	if err := r.queues.AddQueue(ctx, q); err != nil {
 		log.Error(err, "Failed to add queue to system")
 	}
-	return false
+	return true
 }
 
 func (r *QueueReconciler) Delete(e event.DeleteEvent) bool {
-	q := e.Object.(*kueue.Queue)
+	q, match := e.Object.(*kueue.Queue)
+	if !match {
+		// No need to interact with the cache for other objects.
+		return true
+	}
 	r.log.V(2).Info("Queue delete event", "queue", klog.KObj(q))
 	r.queues.DeleteQueue(q)
-	return false
+	return true
 }
 
 func (r *QueueReconciler) Update(e event.UpdateEvent) bool {
-	q := e.ObjectNew.(*kueue.Queue)
+	q, match := e.ObjectNew.(*kueue.Queue)
+	if !match {
+		// No need to interact with the cache for other objects.
+		return true
+	}
 	log := r.log.WithValues("queue", klog.KObj(q))
 	log.V(2).Info("Queue update event")
 	if err := r.queues.UpdateQueue(q); err != nil {
 		log.Error(err, "Failed to update queue in system")
 	}
-	return false
+	return true
 }
 
 func (r *QueueReconciler) Generic(e event.GenericEvent) bool {
 	r.log.V(3).Info("Ignore generic event", "obj", klog.KObj(e.Object), "kind", e.Object.GetObjectKind().GroupVersionKind())
-	return false
+	return true
+}
+
+type queueqWorkloadHandler struct{}
+
+func (h *queueqWorkloadHandler) Create(e event.CreateEvent, q workqueue.RateLimitingInterface) {
+	qw := e.Object.(*kueue.QueuedWorkload)
+	// Check that the wl has been assigned first
+	if qw.Spec.AssignedCapacity != "" {
+		q.Add(requestForQueueStatus(qw))
+	}
+}
+
+func (h *queueqWorkloadHandler) Update(e event.UpdateEvent, q workqueue.RateLimitingInterface) {
+	oldQW := e.ObjectOld.(*kueue.QueuedWorkload)
+	newQW := e.ObjectOld.(*kueue.QueuedWorkload)
+	// If a workload transitioned from pending to assigned or finished, we need to Renconcile in all cases.
+	if newQW.Spec.AssignedCapacity != "" && workloadStatus(newQW) != workloadStatus(oldQW) {
+		q.Add(requestForQueueStatus(newQW))
+	}
+	// If a workload switched queues, we need to update both queues.
+	if newQW.Spec.QueueName != oldQW.Spec.QueueName {
+		q.Add(requestForQueueStatus(oldQW))
+		q.Add(requestForQueueStatus(newQW))
+	}
```

Comment:
I think something like this would be more accurate.

```go
	if newQW.Spec.AssignedCapacity == "" || oldQW.Spec.AssignedCapacity == "" {
		q.Add(requestForQueueStatus(newQW))
		if newQW.Spec.QueueName != oldQW.Spec.QueueName {
			q.Add(requestForQueueStatus(oldQW))
		}
	}
```

But let's just Rencocile.... if nothing changed, there won't be any status update. Reconcile itself is somewhat cheap.

```go
	q.Add(requestForQueueStatus(newQW))
	if newQW.Spec.QueueName != oldQW.Spec.QueueName {
		q.Add(requestForQueueStatus(oldQW))
	}
```